aptos-move/framework/aptos-framework/sources/nonce_validation.move

```diff
@@ -0,0 +1,236 @@
+            nonce_to_exp_time_map: big_ordered_map::new_with_config(0, 0, true),
+        };
+
+        // Initiating big ordered maps with 5 pre-allocated storage slots.
+        // (expiration time, address, nonce) is together 48 bytes.
+        // A 4 KB storage slot can store 80+ such tuples.
+        // The 5 slots should be more than enough for the current use case.
+        bucket.nonces_ordered_by_exp_time.allocate_spare_slots(5);
+        bucket.nonce_to_exp_time_map.allocate_spare_slots(5);
+        bucket
+    }
+
+    // This method is used to prefill the nonce_table with empty buckets one by one.
+    public entry fun add_nonce_buckets(count: u64) acquires NonceHistory {
+        assert!(exists<NonceHistory>(@aptos_framework), error::invalid_state(E_NONCE_HISTORY_DOES_NOT_EXIST));
+        let nonce_history = &mut NonceHistory[@aptos_framework];
+        for (i in 0..count) {
+            if (nonce_history.next_key <= NUM_BUCKETS) {
+                if (!nonce_history.nonce_table.contains(nonce_history.next_key)) {
+                    nonce_history.nonce_table.add(
+                        nonce_history.next_key,
+                        empty_bucket()
+                    );
+                };
+                nonce_history.next_key = nonce_history.next_key + 1;
+            }
+        }
+    }
+
+    // Returns true if the input (address, nonce) pair doesn't exist in the nonce history, and inserted into nonce history successfully.
+    // Returns false if the input (address, nonce) pair already exists in the nonce history.
+    public(friend) fun check_and_insert_nonce(
+        sender_address: address,
+        nonce: u64,
+        txn_expiration_time: u64,
+    ): bool acquires NonceHistory {
+        assert!(exists<NonceHistory>(@aptos_framework), error::invalid_state(E_NONCE_HISTORY_DOES_NOT_EXIST));
+        let nonce_history = &mut NonceHistory[@aptos_framework];
+        let nonce_key = NonceKey {
+            sender_address,
+            nonce,
+        };
+        let nonce_key_with_exp_time = NonceKeyWithExpTime {
+            txn_expiration_time,
+            sender_address,
+            nonce,
+        };
+        let bucket_index = sip_hash_from_value(&nonce_key) % NUM_BUCKETS;
+        let current_time = timestamp::now_seconds();
+        if (!nonce_history.nonce_table.contains(bucket_index)) {
+            nonce_history.nonce_table.add(
+                bucket_index,
+                empty_bucket()
+            );
+        };
+        let bucket = table::borrow_mut(&mut nonce_history.nonce_table, bucket_index);
+
+        let existing_exp_time = bucket.nonce_to_exp_time_map.get_cloned(&nonce_key);
+        if (existing_exp_time.is_some()) {
+            let existing_exp_time = existing_exp_time.extract();
+
+            // If the existing (address, nonce) pair has not expired, return false.
+            if (existing_exp_time >= current_time) {
```

Comment:
don't we expire things if they are MAX_EXPIRATION_TIME past expiration?

Reply:
We make sure the expiration time <= insertion time + MAX_EXPIRATION_TIME when inserting the nonce.
We can garbage collect after the nonce's expiration time passes.